ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -483,22 +488,178 @@ public void hardDeleteTest() throws InterruptedException, IOException, StoreExce
     index.hardDeleter.preLogFlush();
     index.hardDeleter.postLogFlush();
-    long expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    assertEquals("Hard delete should not have progressed", expectedProgress, index.hardDeleter.getProgress());
+
+    // pause should be idempotent
+    index.hardDeleter.pause();
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    verifyHardDeleterThreadState(Thread.State.WAITING, hardDeleterSleepTimeWhenCaughtUpMs + 1,
+        hardDeleterSleepTimeWhenCaughtUpMs / 10);
+    index.hardDeleter.preLogFlush();
+    index.hardDeleter.postLogFlush();
+    assertEquals("Hard delete should not have progressed", expectedProgress, index.hardDeleter.getProgress());
+
+    index.hardDeleter.resume();
+    assertFalse("Hard deletes should not have been paused ", index.hardDeleter.isPaused());
+    expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // after resuming. hard deletes should progress. Give it some time to hard delete next range
+    waitUntilExpectedProgress(expectedProgress, 1000);
+    verifyEntriesForHardDeletes(idsToDelete, true);
+
+    // check that resume is no-op when pause is not called
+    addPutEntries(4, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    // delete two entries
+    idsToDelete.clear();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
+    index.hardDeleter.resume();
+    assertFalse("Hard deletes should not have been paused ", index.hardDeleter.isPaused());
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // after resuming. hard deletes should progress. Give it some time to hard delete next range
+    waitUntilExpectedProgress(expectedProgress, 1000);
+    verifyEntriesForHardDeletes(idsToDelete, true);
+
+    // pause hard deletes and reload the index to verify that hard deletes has been paused
+
+    index.hardDeleter.pause();
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    verifyHardDeleterThreadState(Thread.State.WAITING, hardDeleterSleepTimeWhenCaughtUpMs + 1,
+        hardDeleterSleepTimeWhenCaughtUpMs / 10);
+    // delete two entries
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.clear();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
```

Comment:
you haven't advanced time - even if hard delete thread is running, it won't have touched anything.